lib/go-tc/deliveryservices.go

```diff
@@ -192,18 +194,416 @@ type DeliveryServiceFieldsV31 struct {
+	// HTTP spec) and their values (separated by a colon as per the HTTP spec)
+	// which will be sent by Traffic Router in HTTP responses to client requests
+	// for this Delivery Service's content. This has no effect on DNS-routed or
+	// un-routed Delivery Service Types.
+	TRResponseHeaders *string `json:"trResponseHeaders"`
+	// TRRequestHeaders is an "array" of HTTP headers which should be logged
+	// from client HTTP requests for this Delivery Service's content by Traffic
+	// Router, separated by newlines. This has no effect on DNS-routed or
+	// un-routed Delivery Service Types.
+	TRRequestHeaders *string `json:"trRequestHeaders"`
+	// Type describes how content is routed and cached for this Delivery Service
+	// as well as what other properties have any meaning.
+	Type *DSType `json:"type"`
+	// TypeID is an integral, unique identifier for the Tenant to which the
+	// Delivery Service belongs.
+	TypeID *int `json:"typeId" db:"type"`
+	// XMLID is a unique identifier that is also the second lowest-level DNS
+	// label used by the Delivery Service. For example, if a Delivery Service's
+	// content may be requested from video.demo1.mycdn.ciab.test, it may be
+	// inferred that the Delivery Service's XMLID is demo1.
+	XMLID *string `json:"xmlId" db:"xml_id"`
 }
 
 // DeliveryServiceV4 is a Delivery Service as it appears in version 4 of the
 // Traffic Ops API - it always points to the highest minor version in APIv4.
 type DeliveryServiceV4 = DeliveryServiceV40
 
+// These are the TLS Versions known by Apache Traffic Control to exist.
+const (
+	// Deprecated: TLS version 1.0 is known to be insecure.
+	TLSVersion10 = "1.0"
+	// Deprecated: TLS version 1.1 is known to be insecure.
+	TLSVersion11 = "1.1"
+	TLSVersion12 = "1.2"
+	TLSVersion13 = "1.3"
+)
+
+func newerDisallowedMessage(old string, newer []string) string {
+	l := len(newer)
+	if l < 1 {
+		return ""
+	}
+
+	var msg strings.Builder
+	msg.WriteString("old TLS version ")
+	msg.WriteString(old)
+	msg.WriteString(" is allowed, but newer version")
+	if l > 1 {
+		msg.WriteRune('s')
+	}
+	msg.WriteRune(' ')
+	msg.WriteString(newer[0])
+	if l > 1 {
+		msg.WriteString(", ")
+		if l > 2 {
+			msg.WriteString(newer[1])
+			msg.WriteString(", and ")
+			msg.WriteString(newer[2])
+		} else {
+			msg.WriteString("and ")
+			msg.WriteString(newer[1])
+		}
+		msg.WriteString(" are ")
+	} else {
+		msg.WriteString(" is ")
+	}
+	msg.WriteString("disallowed; this configuration may be insecure")
+
+	return msg.String()
+}
+
+// TLSVersionsAlerts generates warning-level alerts for the given TLS versions
+// array. It will warn if newer versions are disallowed while older, less
+// secure versions are allowed, or if there are unrecognized versions present.
+//
+// This does NOT verify that the passed TLS versions are _valid_, it ONLY
+// creates warnings based on conditions that are possibly detrimental to CDN
+// operation, but can, in fact, work.
+func TLSVersionsAlerts(vers []string) Alerts {
```

Comment:
Yeah, idk. It's not using any information specific to TO, like a database connection or something, so I stuck it in there. I'm not convinced it's the right place, it just seemed like it _could_ go there. Although maybe the notion of Alerts being associated with a set of TLS Versions is inherently TO-centric.

Reply:
Update: this function (now method) is now fully responsible for generating warnings, since we're now properly checking protocol instead of type there's no database dependency at all. So it made sense to me to make it a method of the DS, but it can still easily go in TO if you think that makes more sense.

Reply:
That's fine -- I'll mark this resolved